homeassistant/components/generic/diagnostics.py

```diff
@@ -16,13 +17,15 @@
     CONF_USERNAME,
 }
 
-# A very similar redact function is in components.sql.  Possible to be made common.
-URL_RE = re.compile("//.*:.*@")
-
 
-def redact_credentials(data: str) -> str:
-    """Redact credentials from string data."""
-    return URL_RE.sub("//****:****@", data)
+# A very similar redact function is in components.sql.  Possible to be made common.
+def redact_url(data: str) -> str:
+    """Redact credentials from string url."""
+    url = yarl.URL(data)
+    if url.user is not None:
```

Comment:
Is it possible to have one with a password but no username?

Reply:
Yeah, I wondered that, in fact I tried a case for both first.

So that url would look like `http://monkey123:@example.com`  Seems almost like an invalid url but I've cecked them separately anyway just in case. b23610d